### ISSUE TITLE
Interoperable font metrics via explicit font metrics overrides

### ISSUE THREAD

**Issue by litherum** [w3c/csswg-drafts#4792]
Migrated from https://github.com/w3c/csswg-drafts/issues/2228#issuecomment-380762748

In font files, there exist at least three different ascent/descent values:
- `ascender` and `descender` from the [`hhea` table](https://docs.microsoft.com/en-us/typography/opentype/spec/hhea)
- `sTypoAscender` and `sTypoDescender` from the [`OS/2` table](https://docs.microsoft.com/en-us/typography/opentype/spec/os2)
- `usWinAscent` and `usWinDescent` from the [`OS/2` table](https://docs.microsoft.com/en-us/typography/opentype/spec/os2)
- maybe others???

Different browsers / OSes use different values. In fact, some even use values that are not present in the font file at all. This leads to content being laid out differently on different browsers / OSes. https://github.com/w3c/csswg-drafts/issues/2228 is an example of one. There have been others.

There have been efforts to try to introduce "interoperable" metrics into CSS (https://github.com/w3c/csswg-drafts/issues/3978 is one, https://github.com/w3c/csswg-drafts/issues/938 is related, there are a few others too), to force browsers to agree on which metrics they use. Many of the existing proposals so far have some downsides:
- They require browsers to parse font files, rather than relying on special-purpose font parsing libraries (such as CoreText). A tenet of the web platform is that browsers should be able to delegate responsibility to helper libraries.
- They cause all text to be laid out differently, across every website in the entire web. Such a massive change is too scary from a product development standpoint.
- Any approach which is opt-in for specific cases causes text metrics to not match throughout the page. This can lead to inconsistent spacing and poor typography / webpage design.

However, interoperable font metrics _are_ desirable for many page authors. Here's a compromise proposal that (I believe) solves all the above problems:

What if we add a new descriptor into `@font-face` to let the CSS author override the metrics of the font? Something like this:

```CSS
@font-face {
    font-family: "WebFont";
    src: ...;
    ascent-override: 80%;
    descent-override: 20%;
}
```

Those percentage values are resolved by multiplying by the font size (em box). Negative values are illegal. The browser doesn't have to parse any font files, all browsers will agree on the resolved value, and all text in all situations drawn with that font will use these values.

There's already precedent for this: the `font-feature-values` and `font-variation-values` descriptors in `@font-face` overrides the defaults present in the font file.

**Comment by kojiishi** [comment]
Thank you for writing this up! /cc @kojiishi @drott

**Comment by Crissov** [comment]
The downside of any solution that relies on `@font-face` is that it does not work for local fonts unless authors reference them with `local()`, which they hardly do in my experience. Please correct me if Iʼm wrong.

**Comment by faceless2** [comment]
> (other proposals) cause all text to be laid out differently, across every website in the entire web. Such a massive change is too scary from a product development standpoint.

I suspect not applying this to local (or even generic) fonts might be entirely intentional.

It's a good compromise idea - I still think we need `leading-override` too, as OpenType has at least `hhea.LineGap` and `OS/2.sTypoLineGap` defining this.

There are other metrics to consider. Both x-height and cap-height are both in the OpenType PCLT table, and there are the baselines too (in the BASE table). All of these are stored only in one place in the font, so if this issue is purely about removing ambiguity and not about overriding metrics, then I'm not sure whether overriding these is important.

**Comment by tobireif** [comment]
I really appreciate any and all efforts that have the goal of fixing https://github.com/w3c/csswg-drafts/issues/2228 .

This here ticket wouldn't solve my use-case in a way that's convenient enough - I'd first have to figure out which percentage values to fill in.

A good solution would allow me (and other web developers) to ensure that the described cross-OS-differences won't happen (in modern standards-based browsers) - in a convenient and quick way, eg `glyph-positioning: glyph-bounding-box` (the exact names are up to you / the W3C 😀)

Chrome on MacOS:
![41592189-1b1b8ee4-73bc-11e8-9c21-9ff43ac2167f](https://user-images.githubusercontent.com/80411/74832699-25664000-5318-11ea-8034-97a85873d537.png)
Chrome on Windows:
![41592204-25a82ae8-73bc-11e8-818d-d64216d9c480](https://user-images.githubusercontent.com/80411/74832715-2e571180-5318-11ea-9cc9-f42a2698267a.png)

**Comment by tobireif** [comment]
> They cause all text to be laid out differently, across every website in the entire web. Such a massive
> change is too scary from a product development standpoint.

Any solution should be opt-in.

> Any approach which is opt-in for specific cases causes text metrics to not match throughout the
> page. This can lead to inconsistent spacing and poor typography / webpage design.

That's the responsibility of the web developer. They can apply it to the whole page or only to portions of it - it's up to them. I think that typically the property would be used for eg a large heading and that's it - the rest of the page would be smaller text (where the issue is eg negligible and which might not use the same font as the heading). In any case - it's the responsibility of the web developer to ensure consistent spacing and good typography etc. (That's the exact reason why I had posted #2228  😀 because I was trying to do just that.)

**Comment by tobireif** [comment]
If your proposal will the the only/last offer by the W3C for solving the issue(s) described in #2228 , then I guess I should accept it (towards then closing that ticket) - if it solves the issue(s) described in #2228 completely, and although it requires a bit more work than I had wished for.

**Comment by khaledhosny** [comment]
One can already override font provided line spacing with CSS `line-height` and it does not require any font parsing either, what additional benefit one would get from the proposed `@font-face` overrides?

**Comment by jfkthame** [comment]
Setting `line-height` is not sufficent -- within a given line-height, where does the baseline of the text go? This depends on the relative ascent & descent metrics.

**Comment by faceless2** [comment]
Yes, you can set a specific line-height every time you use that font and get consistent cross-browser results for line-spacing.

But it seems if you're overriding ascender/descender, on which the default value of line-height depends, you might also want to set the default line-height as well. It's not necessary, but it will save you specifying `line-height` everywhere you use the font. If the stylesheet defining the font is being imported by someone not familiar with the modified metrics, this is probably quite important.

**Comment by jfkthame** [comment]
It seems to me we could reasonably use CSS properties (not just font-face descriptors) here: something like `font-ascent`, `font-descent` and `font-linegap`, all with initial values of `auto` (meaning the browser derives metrics from the font or from system APIs or whatever), but with authors having the possibility to use percentages (to be resolved based on em size) instead.

`normal` line height would then be based on the used values of these metrics.

Then it might *also* be useful to have these as font-face descriptors, which would allow webfonts to override the result of `auto`, while leaving system fonts to do their normal thing. So authors could choose to take control just for a specific web font (using the descriptors), or to override the browser's `auto` behavior for arbitrary content in any font (using the properties). All this would be opt-in for authors, so there's no risk to existing content.

An author could then choose to do something like

    :root { font-ascent: 80%; font-descent: 20%; font-linegap: 10%; }

and expect to get consistent line-spacing behavior across browsers and platforms -- with the understanding that the result may not be the *same* as any given browser currently gives by default, but it's entirely under their control.

I suppose the metrics properties (not descriptors) should probably accept absolute lengths, calc expressions, etc. for completeness, although em-based percentages would usually be the most sensible thing to use.

The final piece would be to add a `from-font` value to each of these properties (distinct from `auto`), and define exactly what it means in terms of metrics from an OpenType font file; I would suggest the `sTypo*` values would be the appropriate choice. (This might require browsers to peek into font files, if system APIs don't provide the required metrics, so may be less convenient to implement.)

**Comment by tobireif** [comment]
@khaledhosny wrote:
> One can already override font provided line spacing with CSS `line-height`

`line-height:normal` does not help, see https://github.com/w3c/csswg-drafts/issues/2228#issuecomment-366310641

Chrome on MacOS:
![36321604-9e035454-134a-11e8-9537-55405fe412cc](https://user-images.githubusercontent.com/80411/74859630-a7ba2880-5347-11ea-8087-0a815b5bba63.png)

Chrome on Windows:
![36321633-b52e3cd4-134a-11e8-8526-cadf5e99e0c6](https://user-images.githubusercontent.com/80411/74859710-c0c2d980-5347-11ea-832d-53bc8a7f8272.png)

`line-height: 1` does not help either, see https://github.com/w3c/csswg-drafts/issues/2228#issuecomment-504767148

@faceless2 wrote:
> Yes, you can set a specific line-height every time you use that font and get consistent
> cross-browser results for line-spacing.

The issue is not line-spacing 😀The issue is shown in the above screenshots. When it appears it appears even for a single line or word, on one OS vs another (more at #2228 ).

@jfkthame wrote:
> Setting line-height is not sufficent

True!

**Comment by litherum** [comment]
@jfkthame 

> The final piece would be to add a from-font value to each of these properties (distinct from auto), and define exactly what it means in terms of metrics from an OpenType font file

I believe this fails the “browsers having to parse font files” test.

**Comment by litherum** [comment]
@jfkthame

> It seems to me we could reasonably use CSS properties (not just font-face descriptors) here

Just to make sure I understand, would these properties apply to the block container? And would they override any/all ascent and descent values for every inline descendent?

**Comment by jfkthame** [comment]
> > The final piece would be to add a from-font value to each of these properties (distinct from auto), and define exactly what it means in terms of metrics from an OpenType font file
> 
> I believe this fails the “browsers having to parse font files” test.

Yes - I figured you probably wouldn't be keen on this value (though we have `from-font` for a couple of the text-decoration properties; doesn't that require browsers to parse font files, too?) Anyhow, I see it as an optional part of the overall suggestion; I think it'd be logical to include a `from-font` value, but it's not required in order for the properties to be useful.

> Just to make sure I understand, would these properties apply to the block container? Non-inherited? And would they override any/all ascent and descent values for every inline descendent?

I haven't really thought about how inheritance etc ought to work... the suggestion that an author might write

    :root { font-ascent: 80%; font-descent: 20%; font-linegap: 10%; }

was based on the idea that these properties would inherit (as percentages, to be resolved for each element against its used `font-size`), but maybe that's not the right approach.

Just throwing ideas out, really, and happy to hear ideas on how it would best work (or why it wouldn't).

**Comment by drott** [comment]
If we allow custom overrides, values, percentages, etc. I'd prefer to not have those placed as descriptors in `@font-face` but rather as new properties. I'd also be very supportive of a from-font value. I'd be even more supportive of not allowing custom value overrides, but only some form of keyword to indicate: "use sTypo* only" - in which case I'd be okay with having that as a descriptor in the `@font-face` header.

Background for not placing overrides in `@font-face` are issues https://github.com/w3c/csswg-drafts/issues/2531 as well as my proposal in https://github.com/w3c/csswg-drafts/issues/4358. Placing overrides in @font-face is not supported in Chrome currently and architecturally difficult. We removed `font-variant` in the spec as an override, and I would be happy if we could remove other overrides there as well. 

@litherum if I am not mistaken in skimming through the code, WebKit [already has code for parsing several OpenType fields manually](https://trac.webkit.org/browser/webkit/trunk/Source/WebCore/platform/graphics/opentype) (for `math` table, vertical layout metrics), as well as even for [sTypoMetrics](https://trac.webkit.org/browser/webkit/trunk/Source/WebCore/platform/graphics/opentype/OpenTypeCG.cpp#L57) (but this may not be used when running on Mac OS). I do understand your preference to have CoreText handle metrics parsing, but wouldn't this be an acceptable case of accessing this data to enable a `from-font` keyword? Or alternatively, since you work closely with your CoreText colleagues, couldn't they expose the `sTypo*` metrics for you?

**Comment by fantasai** [comment]
@jfkthame I would prefer to discuss properties (rather than @font-face descriptors) in a separate issue, as it has rather different implications and considerations than an @font-face descriptor. Also for the record, I don't think it's a good idea. Where to place the ascent/descent metrics is very font-dependent, and applying numbers fine-tuned to one font to another is liable to cause overlapping text. As for handling line-gaps and switching out how we measure the height of text, that's a css-inline issue, not a font issue, and is being handled there. [ in reply to https://github.com/w3c/csswg-drafts/issues/4792#issuecomment-588341590 ]

**Comment by chrishtr** [comment]
There is another use case that would I think be solved by the proposal in this issue: matching fallback font metrics to emulate a web font, so as to minimize layout shift during page load. See [here](https://meowni.ca/font-style-matcher/) for one example; this can currently be combined with font face observation [script libraries](https://github.com/bramstein/fontfaceobserver) to dynamically update styles to remove the overrides. If we also add a descriptor such as:

```
letter-spacing-override: 1.1
```

(Which would make spacing 10% bigger) then we can also solve that use case, and avoid the use of these scripts. Using these scripts effectively is quite difficult since the script would have to be loaded extremely early in the page lifecycle. For that reason I think most sites don't use these libraries, even though they might improve page load UX.

**Comment by litherum** [comment]
See: https://github.com/w3c/csswg-drafts/issues/2228#issuecomment-631657902

**Comment by litherum** [comment]
@chrishtr There's a issue https://github.com/w3c/csswg-drafts/issues/450 trying to solve this use case from another angle.

**Comment by xiaochengh** [comment]
From the Chrome side, we'd like to +1 to `ascent-override` and `descent-override` as how @litherum proposed. In addition, we also propose:
- `line-gap-override: <percentage>` to override line gap. It works in the same way as `ascent-override` and `descent-override`
-  `advance-override: <number>` to add additional advance to characters. It's the same as how @chrishtr [proposed](https://github.com/w3c/csswg-drafts/issues/4792#issuecomment-663264158), just renamed for better clarity

We've [prototyped](https://groups.google.com/a/chromium.org/d/msg/blink-dev/h21QrfPrkl4/ff_dZxYcCQAJ) these descriptors in Blink M86 behind a flag. The [implementation](https://chromium-review.googlesource.com/q/hashtag:%22font-metrics-override%22+(status:open%20OR%20status:merged)) isn't too complicated.

---

Besides interop, we are motivated by the possibility to reduce [cumulative layout shift](https://web.dev/cls/) caused by web fonts. This has also been discussed in #450.

We analyzed a set of AMP pages with CLS issues, and [found that](https://docs.google.com/document/d/1PW-5ML5hOZw7GczOargelPo6_8Zkuk2DXtgfOtJ59Eo/edit?ts=5f4ee491#bookmark=id.q0aftwc4vo6n):
- Font-caused CLS is a pretty common issue. In many cases, fonts on their own already contribute >0.1 CLS, exceeding the "good" threshold
- The new descriptors can eliminate most of the font-caused CLS when configured correctly
  - One one page, it reduced CLS from 0.93 to 0.64, while the "raw" CLS is 0.56 if we disable web fonts

---

One particular advantage of these descriptors is easy adoptation, especially when using 3rd party fonts from a font provider. Font providers may set the correct values of these descriptors in the stylesheets they provide (which could otherwise be pretty cumbersome for individual web authors), while web authors only need to change one line of code in their `font-family` CSS property. For example:

Font provider side:

```
font.provider/fonts.css:

@font-face {
  font-family: foo;
  src: url("http://example.com/foo.woff");
  font-weight: 400;
  ...
}

@font-face {
  font-family: fallback-to-arial;
  src: local(Arial), url("http://example.com/arial.woff");
  font-weight: 400;
  advance-override : 0.1;
  ascent-override: 60%;
  descent-override: 40%;
  line-gap-override: 0%;
}
```

Author side:

```
main.html:

<link rel="stylesheet" href="font.provider/fonts.css">
<style>
 body {
   font-family: foo, fallback-to-arial;
   font-weight: 400;
  }
</style>
```

---

In summary, we found this proposal straightforward enough in terms of both the spec and the implementation, and yet allows easy opt-in. We've also tried some other approaches (e.g., changing how font fallback stack works), but couldn't find one with all these advantages. So we think this should be the best way to go.

(A Google Docs version of our proposal is at https://bit.ly/39qATQ4)

**Comment by tobireif** [comment]
@xiaochengh Will this also fix #2228?

**Comment by chrishtr** [comment]
In addition to issue 450, there is issue 126, which is also about fallback fonts causing layout jumping / Cumulative Layout Shift (CLS) impact.

@tobireif : yes I think it will solve issue 2228 as well.

**Comment by litherum** [comment]
`line-gap-override: <percentage>` makes sense to me.

I think `advance-override` needs some more time to bake, since it isn't actually adjusting font metrics - it's adjusting the `letter-spacing` CSS property. The CSS spec already today says that `letter-spacing` isn't just adding a constant to the advance of each glyph, but instead understands where inline elements end and avoids adding extra space at the end of runs. The behavior specified in the spec today is desirable - simply adding a constant to each glyphs' advance leads to bad typography. Correctly implementing this more complicated spacing algorithm would be tricky to do if one font in the font-family list has advance-override set to one thing but another one has it set to another thing.

**Comment by xiaochengh** [comment]
Thanks for the feedback!

Regarding `advance-override`: Though it looks very similar to the `letter-spacing` CSS property, they are actually for different purposes. `letter-spacing` is for good typography, while `advance-override` is proposed mainly to reduce CLS when switching from a local fallback font to a web font. Therefore, they should work in different ways, and issues with `letter-spacing` (like #1518) do not necessarily apply here.

For the goal of reducing CLS, we want to match the text widths of the fallback font and web font, which requires `advance-override` to adjust the advance of each glyph†, not just the spacing between glyphs (so, we do want extra space at the end of runs). In this way, the width difference between fallback font and web font can be minimized. In the best case, if we are using monospace, we can even perfectly match the widths and get 0 CLS by setting `advance-override: x - y`, where `x` and `y` are the character widths of the web font and fallback, respectively; If we override the spacing between glyphs only, the adjusted with is always off by `x - y`.

Below are some examples where overriding the spacing leads to larger CLS when switching font:
- In a table where the column widths depend on text widths. We want to match the text width as much as possible to reduce the horizontal shifting of the cells
- In a line with multiple fonts, e.g. in [example 23 in the spec](https://drafts.csswg.org/css-text-3/#example-36282bb9:~:text=A%20given%20value%20of%20letter%2Dspacing%20only,element%20for%20which%20it%20is%20specified%3A), where we want spacing applied to the 2nd "b"
- In a line with mixed text and non-text inline elements
- When line wraps, the advance of the last character (very slightly) affects the total width and therefore affects where line breaks

Btw, this is also the reason we renamed it to `advance-override`: to emphasize what it actually overrides.

†  At a more detailed level ([illustration](https://www.freetype.org/freetype2/docs/glyphs/glyphs-3.html)), advance = left side bearing (lsb) + glyph width + right side bearing (rsb). To match text width, `advance-override` should adjust lsb and rsb of every glyph, which is essentially adjusting the advance of every glyph.

**Comment by fantasai** [comment]
Agree with @litherum wrt concerns about `advance-override`.

Wanted to note that we have also talked about metrics for superscript/subscript, so that would be something like:
- `superscript-shift-override` & `subscript-shift-override`
- `superscript-size-override` & `subscript-size-override`

**Comment by tabatkins** [comment]
Hm, `advance-override` would only work reasonably for monospace fonts, right? On a variable-width font, it would set the advance for all characters to the override value, effectively turning it into monospace, which seems bad.

(Solving the "fallback screws up your monospace grid" problem is important on its own, of course, but being able to solve the "tweak a fallback font so that it ends up very close to the primary font" for variable-widths would also be great.)

**Comment by tabatkins** [comment]
But I suppose the variable-width case *does* need something more like letter-spacing, as Myles indicates. So it'll need to be something different anyway.

But yeah, for the monospace case we definitely do *not* want to act like letter-spacing, we absolutely want to affect the actual advance, so this seems like the right approach for monospace fonts. Just need to make sure the name isn't going to be confusing with whatever we do for variable-width things later. (I think it's probably okay, as the name will probably invoke `letter-spacing` somehow.)

**Comment by faceless2** [comment]
Not directly relating to the proposed descriptors, but the general approach opens the door for a lot of other descriptors which might be useful, not least being able to override (or even invent) baselines positions, i.e. being able to set the a "hebrew-top" baseline or other script-specific baselines as suggested in https://github.com/w3c/csswg-drafts/issues/5244

**Comment by tabatkins** [comment]
> Hm, advance-override would only work reasonably for monospace fonts, right? On a variable-width font, it would set the advance for all characters to the override value, effectively turning it into monospace, which seems bad.

Ah, I was misled by the name; as currently specified, it adds a value to the existing advance of each character. So to get a fallback monospace to match the primary monospace, you set it to the difference in their two widths. The name should be changed, or the meaning should match the name. ^_^

This *would* then work for variable-width fonts, but as Elika said in the call and Myles suggested in earlier comments, handling variable-width well probably wants a smarter solution: `l` and `w` probably want different amounts of advance adjustment; it probably wants to act like letter-spacing and not add space at the end of a span; etc.

**Comment by tobireif** [comment]
Please make sure that implementers will be required to make sure that your new CSS addition will be a reliable a way for web developers of working around the issue(s) described in #2228 .
Also see for example https://bugs.chromium.org/p/chromium/issues/detail?id=1116349 and https://bugs.chromium.org/p/chromium/issues/detail?id=800693 .

**Comment by xiaochengh** [comment]
I just put up #5521 to add `ascent-override`, `descent-override` and `line-gap-override` to the spec. Could anyone review it?

**Comment by svgeesus** [comment]
> I just put up #5521 to add ascent-override, descent-override and line-gap-override to the spec. Could anyone review it?

The PR was reviewed and has been merged. You can [see it in the spec here](https://drafts.csswg.org/css-fonts-4/#font-metrics-override-desc).

Well, now we need WPT tests :)

**Comment by xiaochengh** [comment]
I guess we also need to modify the JS `FontFace` interface to include these new descriptors?

**Comment by svgeesus** [comment]
> I guess we also need to modify the JS FontFace interface to include these new descriptors?

Yes, indeed

**Comment by litherum** [comment]
> > I guess we also need to modify the JS FontFace interface to include these new descriptors?
> 
> Yes, indeed

Does this need a separate WG resolution? I hope not.

**Comment by astearns** [comment]
No additional resolution needed. It follows directly from the existing resolution.

**Comment by xiaochengh** [comment]
Could anyone review #5538? Just a trivial patch adding attributes to `FontFace` for the overrides.

**Comment by yisibl** [comment]
![2020-09-23 16_18_47](https://user-images.githubusercontent.com/2784308/93990704-0522c600-fdbe-11ea-9fcf-7f6fe2fe0a29.gif)
How to solve the punctuation marks, I tried to use unicode-range to cover the punctuation marks individually, and found that line-gap-override will take effect for the entire font. Advance-override can only control the right width of characters. Is this a bug in Chrome?

Demo: https://codepen.io/yisi/pen/xxVmzOM

```css
@font-face {
  font-family: 'STSongti-SC-Regular-punctuation';
  advance-override: -0.4; 
  src: local(STSongti-SC-Regular);
  unicode-range: U+b7;
}
```

![image](https://user-images.githubusercontent.com/2784308/93994044-07871f00-fdc2-11ea-8e90-d6e0e7bb9ab7.png)

**Comment by xiaochengh** [comment]
@yisibl Thanks for trying the prototype 😄
- I'm not sure what you mean by "`line-gap-override` will take effect for the entire font". If you believe it's a Chrome bug, could you file a bug at crbug.com instead?
- We haven't reached any resolution on `advance-override` should work yet, so Chrome's current behavior is just arbitrary and can be changed at any time. Please use #5533 if you want to join the spec discussion

**Comment by tobireif** [comment]
Thanks all! Wrote a post: https://tobireif.com/posts/ensuring_the_correct_vertical_position_of_large_text/

**Comment by chrishtr** [comment]
Closing. Further discussion can happen on #126 and #5533.
